[Feature] pairwise phase consistency

TRuikes opened the pull request:
Hi! 
In our lab we are using 'pairwise phase consistency' as a metric to measure spike-LFP interactions. The conventional method 'Phase Locking Value' can give faulty output if there are few spikes, and PPC improves on this. ([function as implemented by authors](https://www.fieldtriptoolbox.org/reference/ft_connectivity_ppc/). In short you compute the phase of an lfp during spike times (spike_triggered_phase) and then compute the angular distance between each pair of phases.

The output for the function `phase_analysis.spike_triggered_phase` can be directly fed into this function. While PPC can also be used to compute synchony between two EEG/LFP signals, I think there are allready other (better) measures for this and we can leave it at this simple implementation for now.

It can be a nice addition, let me know if you want to add this and if so what you would like me to change/add. In addition I suggest a tutorial on 'spike-field' functionality of elephant, briefly describing basic LFP processing steps (filtering, hilbert transform) and possible metrics/functions (spike_triggered_phase, sta.spike_field_coherence, sta.spike_triggered_average (<-that's PLV)).

Commit:
Draft version of 'pairwise phase consistency'

pep8speaks commented:
Hello @TRuikes! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`elephant/phase_analysis.py`](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py):

> [Line 189:80](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L189): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 79 characters)
> [Line 192:80](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L192): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 79 characters)
> [Line 197:80](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L197): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 208:80](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L208): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 213:80](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L213): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 79 characters)
> [Line 222:80](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L222): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 224:80](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L224): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 79 characters)
> [Line 231:80](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L231): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 233:80](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L233): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 235:80](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L235): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 79 characters)
> [Line 246:1](https://github.com/NeuralEnsemble/elephant/blob/dd96e0a20d49ad2c8e0ebc1b8d7629bbf798f62f/elephant/phase_analysis.py#L246): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

dizcza committed:
bibtex citation, doc style

dizcza commented:
Hi Thijs,
we're currently advancing phase statistics (see #385), and we do have PPC on the to-do list, thanks for bringing this!
We're also porting the documentation to the new BibTex citation documentation format, so I've updated your docs & reference in https://github.com/TRuikes/elephant/pull/1. You can merge my PR here. I still don't see you in the authors list of Elephant!

The tests are missing. If you're lazy to write them, I can come up with something, though I'm not an expert in this field.

TRuikes committed:
Merge pull request #1 from INM-6/feature/ppc

bibtex citation, doc style

TRuikes commented:
Allright, thanks for the quick response! I'll think of some tests (wasn't too lazy, just didn't want spend time before knowing if you were going to use it ;) ) and add myself to authors.

The authors made impoved versions of PPC (this one they call 'ppc0' and then there are two more, I think). Long story short, when computing PPC between pairs, they dont want to draw samples from the same trial (their full reasoning [written here](http://europepmc.org/article/PMC/3394239#impact), due to spiking properties of a neuron). To compute this I'd have to give information on trial nr per phase. I think I'll add an optional keyword `trial_ids` whith same dimensions as `phases` with a number per trial. Then another keyword to select `ppc0` or `ppc1`.

I'm going to work on this next week again (monday), so I'll push something around then.

dizcza commented:
> I think we don't need to multiply by 2. The matrix dot_prod is completely filled, so each pair 'i','j' is computed twice.

Yes, you're right, I didn't think of that.

Regarding your previous comment about ppc1, if you think that it could take much time or you're unsure how to implement this additional feature that potentially (!) improves the estimate, it's better to come up with another pull request that addresses the improvement after this PR is merged.

Commit:
Consider each entry in the `phases` input as a 'trial' from the same spiketrain

dizcza commented on elephant/phase_analysis.py:
```diff
@@ -235,27 +244,35 @@ def pairwise_phase_consistency(phases):
+    # Compute the distance between each pair of phases using dot product
+    # Optimize computation time using array multiplications instead of for
+    # loops
+    p_cos_2d = np.tile(np.cos(phase_array), reps=(n_trials, 1))  # TODO: optimize memory usage
```
right, `numpy.broadcast_to` is recommended since it does not consume memory as tile does.

TRuikes replied:
I was thinking about specifying datatype, using np.float32 would allready reduce half of the memory used. I noticed when trying this on large spiketrains that it consumes too much RAM (nspikes^2). I don't know if broadcast_to will use less memory than tile, I'm not planning on testing it, unless you think it's worth the time.

dizcza replied:
According to numpy's doc, broadcast_to should use only N memory instead of N^2, as in tile.
Because it does not allocate memory but rather returns a read-only view with the given shape. And it's everywhere stated through the numpy doc that broadcast_to should be preferred for exactly this reason - memory.

dizcza replied:
Therefore, we can use the datatype returned by `cos`, which is double float by default.

TRuikes replied:
ah I see, since its a view it wont actually construct a new array. And would we save memory if we add `dtype=np.float32` to `np.multiply` in the following line? Because the `dot_prod` is going to be a new matrix of size N^2 (not a view like cos).

dizcza replied:
ah, yes, so we do need to have N^2 after all.
Then yes, I'd cast everything to float32

dizcza commented on elephant/phase_analysis.py:
```diff
@@ -235,27 +244,35 @@ def pairwise_phase_consistency(phases):
+    if method not in ['ppc0']:
+        raise ValueError('For method choose out of: ["ppc0"]')
+
+    phase_array = np.hstack(phases)
```
are you sure that all phases have the same length?

TRuikes replied:
I'm sure all phases _do not_ have the same length :p. I implemented this method (slightly against your earlier advice) after I discovered that simply putting in a whole spiketrain for a 2h session takes too much RAM. As I had to cut up my spiketrain anyway, it felt like it made sense to do so in trials. As such the default input for the function will be (if you agree) a list of arrays, where each array is a trial of size `n_phases`. For the default version of PPC we can simply append all these phases, and compute the consistency, and I'll leave it at this for this PR.

Then in a future PR we can easily extend this function with PPC1, by selecting indices of the input phases, such that we only average between pairs of phases between trials (i.e. PPC1).

Commit:
updated phase_analysis.py:
-changed np.tile to np.broadcast_to
-added dtype=np.float32

added tests in test_phase_analysis.py

TRuikes commented on elephant/test/test_phase_analysis.py:
```diff
@@ -202,5 +202,131 @@ def test_regression_269(self):
+class PairwisePhaseConsistencyTestCase(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):  # Note: using setUp makes the class call this
```
Do you prefer `setUp` over `setUpClass`?

dizcza replied:
They serve different use cases: `setUp` is run at each test while `setUpClass` is run only once at the test class initialization. Since you're initializing read-only values only once, `setUpClass` is preferred for the sake of runtime. If the initialization is fast, either way is fine.

dizcza commented:
Let me know when it's ready for review.

TRuikes commented:
It's ready for review! I'm allready testing it on a (large) dataset and will let you know if something comes up.

TRuikes commented:
If I'm supposed to do something please let me know

dizcza committed:
fixed the citation, removed todos, pep8

dizcza committed:
Merge branch 'master' into pull/ppc0

dizcza reviewed (approved):
I created a PR again https://github.com/TRuikes/elephant/pull/2

There was a todo "handle nans". Nans can occur iff the sum is zero and n_trials=1. In such cases, nans are the valid output.

Once the PR is merged, I merge this.

Commit:
Merge branch 'master' into pull/ppc0

# Conflicts:
#	doc/authors.rst
#	doc/bib/elephant.bib
#	doc/reference/phase_analysis.rst
#	elephant/phase_analysis.py

TRuikes committed:
Merge pull request #2 from INM-6/pull/ppc0

Fixed many small things about PPC0

Moritz-Alexander-Kern commented:
Hi @TRuikes ,
To keep the whole thing neatly organized, it seems to make sense to merge this as one PR. 
So I suggest to continue on this after september, otherwise we would have to scatter the PPC1extension and additional tests across multiple PRs.
Nice work so far, looking forward to continue working on this with you.

Moritz-Alexander-Kern committed:
Merge branch 'master' into master
